sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -38,95 +260,208 @@ static kmac_error_t wait_status_bit(uint32_t bit_position, uint32_t bit_value) {
-      break;
+  return kKmacOk;
+}
 
-    default:
-      return kKmacNotImplemented;
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_write_prefix_block(kmac_operation_t operation,
+                                     const uint8_t *func_name,
+                                     const size_t func_name_len,
+                                     const uint8_t *cust_str,
+                                     const size_t cust_str_len) {
+  kmac_error_t err = kKmacOk;
+  // Write to prefix registers
+  if (operation == kKmacOperationCSHAKE) {
+    err =
+        kmac_set_prefix_regs(func_name, func_name_len, cust_str, cust_str_len);
+  } else if (operation == kKmacOperationKMAC) {
+    uint8_t kmac_func_name[4] = {0x4b, 0x4d, 0x41,
+                                 0x43};  // "KMAC" in little endian
+    err = kmac_set_prefix_regs(kmac_func_name, /*func_name_len=*/4, cust_str,
+                               cust_str_len);
+  }
+  if (err != kKmacOk) {
+    return err;
+  }
+  return err;
+}
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_write_key_block(const uint8_t *key, kmac_key_len_t key_len) {
+  if (key == NULL) {
+    return kKmacArgsError;
   }
-  abs_mmio_write32_shadowed(
-      TOP_EARLGREY_KMAC_BASE_ADDR + KMAC_CFG_SHADOWED_REG_OFFSET, cfg_reg);
+
+  uint32_t key_len_reg = KMAC_KEY_LEN_REG_RESVAL;
+  key_len_reg =
+      bitfield_field32_write(key_len_reg, KMAC_KEY_LEN_LEN_FIELD, key_len);
+  abs_mmio_write32(kKmacBaseAddr + KMAC_KEY_LEN_REG_OFFSET, key_len_reg);
+
+  size_t key_len_bytes = key_len_lookup[key_len];
+  uint32_t *word_ptr = (uint32_t *)key;
+  for (size_t i = 0; i < key_len_bytes; i += 4) {
+    abs_mmio_write32(kKmacKeyShare0Addr + i, *word_ptr++);
+    abs_mmio_write32(kKmacKeyShare1Addr + i, 0x00);
+  }
+
   return kKmacOk;
 }
 
 OT_WARN_UNUSED_RESULT
-kmac_error_t kmac_update(const uint8_t *data, size_t data_len, uint8_t *digest,
-                         size_t digest_len) {
+kmac_error_t kmac_update(kmac_operation_t operation, const uint8_t *data,
+                         size_t data_len, uint8_t *digest, size_t digest_len) {
   uint32_t cmd_reg = KMAC_CMD_REG_RESVAL;
   kmac_error_t err;
 
   // Assumption(1): `data_len` > 0 and `digest_len` > 0
   // Assumption(2): data % 4 = 0, i.e. the data ptr is 32-bit aligned
 
-  err = wait_status_bit(KMAC_STATUS_SHA3_IDLE_BIT, 1);
+  // err = wait_status_bit(KMAC_STATUS_SHA3_IDLE_BIT, 1);
   // Issue the start command, so that messages written to MSG_FIFO are forwarded
   // to Keccak
   cmd_reg = bitfield_field32_write(cmd_reg, KMAC_CMD_CMD_FIELD,
                                    KMAC_CMD_CMD_VALUE_START);
-  abs_mmio_write32(TOP_EARLGREY_KMAC_BASE_ADDR + KMAC_CMD_REG_OFFSET, cmd_reg);
+  abs_mmio_write32(kKmacBaseAddr + KMAC_CMD_REG_OFFSET, cmd_reg);
+  err = wait_status_bit(KMAC_STATUS_SHA3_ABSORB_BIT, 1);
+  if (err != kKmacOk) {
+    return err;
+  }
 
   // Begin by writing a word at a time
   uint32_t *word_ptr = (uint32_t *)data;
```

Comment:
This is also a strict aliasing violation. At this level of abstraction, I'm OK with the KMAC driver accepting `data` as a `uint32_t` and still accepting `data_len` in terms of byte length; higher levels up can handle alignment issues if needed.